scipy/spatial/transform/_rotation_cy.pyx

```diff
@@ -0,0 +1,1425 @@
+            d = quat[ind, k] * sign
+        else:
+            a = quat[ind, 3] - quat[ind, j]
+            b = quat[ind, i] + quat[ind, k] * sign
+            c = quat[ind, j] + quat[ind, 3]
+            d = quat[ind, k] * sign - quat[ind, i]
+
+        _get_angles(angles[ind], extrinsic, symmetric, sign, pi / 2, a, b, c, d)
+
+    if degrees:
+        angles = np.rad2deg(angles)
+    return np.asarray(angles)
+
+
+@cython.embedsignature(True)
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def as_davenport(double[:, :] quat, double[:, :] axes, order, bint degrees=False) -> double[:, :]:
+    cdef np.ndarray[double, ndim=2] q = np.asarray(quat)
+
+    cdef bint extrinsic
+    if order in ['e', 'extrinsic']:
+        extrinsic = True
+    elif order in ['i', 'intrinsic']:
+        extrinsic = False
+    else:
+        raise ValueError("order should be 'e'/'extrinsic' for extrinsic "
+                            "sequences or 'i'/'intrinsic' for intrinsic "
+                            "sequences, got {}".format(order))
+
+    if len(axes) != 3:
+        raise ValueError("Expected 3 axes, got {}.".format(len(axes)))
+
+    if axes.ndim != 2 or axes.shape[1] != 3:
+        raise ValueError("Axes must be vectors of length 3.")
+
+    cdef np.ndarray[double, ndim=1] n1, n2, n3
+    n1, n2, n3 = np.asarray(axes)
+
+    # normalize axes
+    n1 = n1 / np.linalg.norm(n1)
+    n2 = n2 / np.linalg.norm(n2)
+    n3 = n3 / np.linalg.norm(n3)
+
+    if np.dot(n1, n2) >= 1e-7:
+        raise ValueError("Consecutive axes must be orthogonal.")
+    if np.dot(n2, n3) >= 1e-7:
+        raise ValueError("Consecutive axes must be orthogonal.")
+
+    angles = np.asarray(_compute_davenport_from_quat(
+            q, n1, n2, n3, extrinsic))
+
+    if degrees:
+        angles = np.rad2deg(angles)
+
+    return angles
+
+
+@cython.embedsignature(True)
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def inv(double[:, :] quat) -> double[:, :]:
+    cdef np.ndarray[double, ndim=2] q_inv = np.array(quat, copy=True)
+    q_inv[:, 0] *= -1
+    q_inv[:, 1] *= -1
+    q_inv[:, 2] *= -1
+    return q_inv
+
+
+@cython.embedsignature(True)
+@_transition_to_rng('random_state', position_num=1)
+def random(num=None, rng=None):
+    rng = check_random_state(rng)
+
+    if num is None:
+        sample = rng.normal(size=4)
+    else:
+        sample = rng.normal(size=(num, 4))
+
+    return sample
+
+
+@cython.embedsignature(True)
+def identity(num: int | None = None) -> double[:, :]:
+    if num is None:
+        return np.array([0, 0, 0, 1])
+    q = np.zeros((num, 4))
+    q[:, 3] = 1
+    return q
+
+
+@cython.embedsignature(True)
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def magnitude(double[:, :] quat) -> double[:, :]:
+    cdef Py_ssize_t num_rotations = quat.shape[0]
+    cdef double[:] angles = _empty1(num_rotations)
+
+    for ind in range(num_rotations):
+        angles[ind] = 2 * atan2(_norm3(quat[ind, :3]), abs(quat[ind, 3]))
```

Comment:
This might be vectorizable

Reply:
I was planning on doing another optimization PR similar to the one for `RigidTransform`. This might be a good fit for it. Some Cython loops are faster than using the `numpy` broadcasting though, so this will have to get benchmarked.